src/main/scala/rocket/Rocket.scala

```diff
@@ -213,10 +213,10 @@ class Rocket(implicit p: Parameters) extends CoreModule()(p)
   val id_amo_rl = id_inst(0)(25)
   val id_fence_next = id_ctrl.fence || id_ctrl.amo && id_amo_rl
   val id_mem_busy = !io.dmem.ordered || io.dmem.req.valid
+  when (!id_mem_busy) { id_reg_fence := false }
   val id_rocc_busy = Bool(usingRoCC) &&
     (io.rocc.busy || ex_reg_valid && ex_ctrl.rocc ||
      mem_reg_valid && mem_ctrl.rocc || wb_reg_valid && wb_ctrl.rocc)
-  id_reg_fence := id_fence_next || id_reg_fence && id_mem_busy
   val id_do_fence = id_rocc_busy && id_ctrl.fence ||
     id_mem_busy && (id_ctrl.amo && id_amo_aq || id_ctrl.fence_i || id_reg_fence && (id_ctrl.mem || id_ctrl.rocc))
 
```

Comment:
The logic still doesn't look correct (from perf point of view):

![id_reg_fence](https://cloud.githubusercontent.com/assets/16438432/24727295/52c7d314-1a0a-11e7-8c39-6984e0b1dfab.png)

At M1 there is an AMO w/ rl in the decode stage and id_reg_fence is set in the next cycle.

At M2 there is a regular mem op in the decode stage but id_reg_fence is still set.